### DISCUSSION TITLE
CLOG 1.5 - Dev Journal

### DISCUSSION THREAD

**Discussion by rabbibotton** [rabbibotton/clog#158] [General]
The goal of 1.5 is to improve the coding experience and workflow, work on some additional custom controls, etc.

**Comment by rabbibotton** [comment]
Added w3.css styled controls to builder.

**Comment by rabbibotton** [comment]
Removed every use of defsetf and replaced with true (setf x) defun's - word of warning CLOS and defsetf do not play well together. This only started showing up with children in different packages overriding defsetf methods and reason only seeing now when we introduced plug-ins.

**Comment by rabbibotton** [comment]
WARNING: Going forward I am going to start making use of plug-ins in the builder (for now just clog-ace). Which means cloning their repos or using UltraLisp to get them for the dev versions of the builder.

**Comment by sabracrolleton** [comment]
Speaking of ace, is there any way to put a gui menu bar at the top of the editor window (not using builder) following along in demo 3? Having the menu bar in a separate window is unappealing from a ui standpoint.

**Reply by rabbibotton** [reply]
It is certainly possible. I have plans to review menus and context menus soon, but you could do easily building the menus from scratch using the same elements I used to make them, divs and buttons and w3- styles.

**Reply by sabracrolleton** [reply]
My problem is not building the menu; my problem is with understanding how to place Ace. Per Ace documentation, I should be able to instantiate it with the html-id of a div. I apparently do not know the right magic words because Ace refuses to play unless I give it the entire body of the win (at which point buttons do not appear, but their content is now a line in the editor window). Clearly the problem is my understanding, but a working example might help me. (or might not and I should go back to sql land where I belong).

**Reply by rabbibotton** [reply]
The main thing is that ace fills it's parent. So take a look at the start-test function in clog-ace.lisp you will see there that that it fills the middle panel. The other option is to immediately after create set positioning to :absolute and give it dimensions.

**Reply by sabracrolleton** [reply]
Thanks. That worked.

**Comment by rabbibotton** [comment]
Events now use clog-ace (don't forget UltraLisp (ql:update-all-dists) or git pull on both clog-ace and clog)

So that means lisp aware editor and highlighting etc. At some point I will expand their mode template (based on textmate) for common-lisp but it is decent as is.

The event drop down shows the status of events if have content or not via squares or filled squares.

I am going to attempt to add functionality to the builder via swank to find symbols ala M-., code completion, etc. we will see.

<img width="1268" alt="Screen Shot 2022-07-12 at 11 50 36 PM" src="https://user-images.githubusercontent.com/11616789/178646872-d730f1f0-6cda-48c7-991b-5c81ecb1298c.png">

**Comment by rabbibotton** [comment]
Basic autocomplete

![Screen Shot 2022-07-13 at 12 09 45 PM](https://user-images.githubusercontent.com/11616789/178780842-e12bbe55-1a25-43f1-ac64-962bfd39ac81.png)

**Comment by rabbibotton** [comment]
Now auto complete ala swank - so full auto complete

![Screen Shot 2022-07-13 at 3 01 05 PM](https://user-images.githubusercontent.com/11616789/178811858-e07ce730-2ecc-41d0-a26b-8c402c2632f7.png)

**Comment by rabbibotton** [comment]
Now a status bar with argument display, working on more :)

<img width="643" alt="Screen Shot 2022-07-14 at 10 33 55 PM" src="https://user-images.githubusercontent.com/11616789/179135689-727f0771-655e-4583-9116-cf51c1bc4bcb.png">

**Comment by rabbibotton** [comment]
Now tool tips on the status bar show documentation

<img width="645" alt="Screen Shot 2022-07-14 at 11 36 26 PM" src="https://user-images.githubusercontent.com/11616789/179145272-77e94993-7cb1-43be-8754-6627bb728aed.png">

**Comment by rabbibotton** [comment]
All the controls on the panel are now contained in the typeahead, when you say choose button-1 it will expand to (button-1 panel).

**Comment by rabbibotton** [comment]
If you launch via emacs then the autodoc function is used to give you the exact same status line as in emacs with defaults for keys. If you launch from sbcl, vscode, etc you get the regular list of arguments.

**Comment by rabbibotton** [comment]
Command-. or Alt-. will act like M-. and bring you in emacs (if running emacs) to the file and location of the definition where your cursor is :)

**Comment by rabbibotton** [comment]
Today made a slight internal change to builder's code rendering, most will not notice it but helps with Lisp custom controls. The builder now insures that the on-setup/on-create events for a control are executed before any other events are run.

**Comment by rabbibotton** [comment]
Builder Tutorial 5 - Custom Controls
https://docs.google.com/document/d/1nYyGVQwEZs5pq1f3GPkWFvwTvZyDp6fFfFTNyfLBdwY

**Comment by rabbibotton** [comment]
Now can load source files of any type to edit in the Builder. All the features of the event editor are there (M-., autocomplete, colors) but you also get full auto complete and color for another 100 languages, error detection for html, JavaScript and CSS as well :)

Try out F1 (ace command bar) and also ctrl/cmd-, (ace config)

(Don't forget pull clog-ace and clog)

<img width="768" alt="Screen Shot 2022-07-19 at 9 28 17 PM" src="https://user-images.githubusercontent.com/11616789/179875921-9fbfe2f6-9dd1-495d-87bf-1d3f30a50034.png">

**Comment by aykaramba** [comment]
Wow.

**Reply by rabbibotton** [reply]
My plan is to have working project panel also so you can select files, panels, pages, etc
That opens the ability to eventually design through point meta areas and click your way to entire apps, web pages, etc and even auto generate things like, clear page and display panel X, hide panel Y, etc

**Comment by rabbibotton** [comment]
New control - Custom HTML Block
This takes whatever HTML you enter and adds it to the panel as one single control. This is great for say pasting in a ton of text or other html that there is no need to have as distinct controls.

**Reply by rabbibotton** [reply]
I fixed issues with changing the html with the html content box and the end of the properties.
I added a button to convert the Custom HTML block to a DIV, once done you can save and reload panel (or copy and paste) and have access to child controls.

**Comment by rabbibotton** [comment]
So now CLOG Builder has REPL window :)

![Screen Shot 2022-07-20 at 1 37 46 PM](https://user-images.githubusercontent.com/11616789/180046960-42a2cd7d-8b41-4106-bb01-d0364f1ca783.png)
![Screen Shot 2022-07-20 at 1 50 13 PM](https://user-images.githubusercontent.com/11616789/180049388-0331ca7e-9c87-4f76-8aa8-80afe2dde755.png)

**Reply by rabbibotton** [reply]
Very important you need to now also git clone clog-terminal or use UltraLisp
